### PR TITLE
[OP]Spatial Transformer Network

### PR DESCRIPTION
**Pull request by tornadomeet**
releated #2175 
paper *Spatial Transformer Networks* : https://arxiv.org/abs/1506.02025
@sxjscience

### CODE REVIEW

**Comment by sxjscience** [comment]
Looks good!

File: tests/python/unittest/test_operator.py [hunk 1]
```diff
@@ -1135,6 +1135,47 @@ def test_flip():
...
+                    args['loc_conv_weight'] = mx.nd.zeros((num_filter, data_shape[1], kernel[0], kernel[1]), ctx=dev)
+                    args['loc_conv_bias'] = mx.nd.zeros((num_filter,), ctx=dev)
+                    args['loc_fc_weight'] = mx.nd.zeros((6, num_filter*data_shape[2]*data_shape[3]), ctx=dev)
+                    args['loc_fc_bias'] = mx.nd.array([0.5, 0, 0, 0, 0.5, 0], ctx=dev)
+                    grad_grad = [mx.nd.zeros(shape, ctx=dev) for shape in arg_shapes]
+                    exe = stn.bind(dev, args=args, args_grad=grad_grad)
+                    exe.forward(is_train=True)
+                    out = exe.outputs[0].asnumpy()
+                    # check forward
+                    reldiff(out, args['data'].asnumpy()[:, :, h/4:h-h/4, w/4:w-w/4]) < 1e-6
```
Comment (sxjscience) [inline tests/python/unittest/test_operator.py]: Do we need to test more general cases? Like other types of `loc`?

**Comment by tornadomeet** [comment]
@sxjscience  yes, every loc symbol your define is supported, you can check the train_mnist.py example

**Comment by tornadomeet** [comment]
@sxjscience  Hello, would you help fix the lint error in ```python/mxnet/initializer.py``` , https://travis-ci.org/dmlc/mxnet/jobs/140714936, i have no idea how to fix it, i just add it like ```_init_bilinear()```

**Comment by tornadomeet** [comment]
another , the python test failed is brought from this pr #2538

**Comment by sxjscience** [comment]
@tornadomeet It seems that `initializer.py` has too many `elif` (Too many branches). We can disable the lint check like https://github.com/dmlc/mxnet/blob/master/python/mxnet/ndarray.py#L740-L748.

**Comment by sxjscience** [comment]
@tornadomeet I think it would be better if we can perform gradient testing for other cases for the non-CuDNN implementations. I can double check this after the PR is merged.

**Comment by tornadomeet** [comment]
@sxjscience  yes, it better do it!  I had done something on it locally, but not very totally. (not provided in this pr, because i changed the c++ code to check the grad of  internal variable:```grid_src``` and ```grid_dst```), at last, i use the mnist example to do that, i found the non-cudnn  stn almost has the same acc with cudnn stn, but not sure that the gradient is absolutely right~

**Comment by tornadomeet** [comment]
@sxjscience it has passed the basic test on travis, https://travis-ci.org/dmlc/mxnet/builds/140744026, the Apple machine failed because the build processing is broken, which also will be failed  for other PRs.

**Comment by tqchen** [comment]
merge this in for now as mac tests current brokes

**Pull request merged**